src/com/sap/piper/tools/JavaArchiveDescriptor.groovy

```diff
@@ -11,30 +11,59 @@ class JavaArchiveDescriptor extends ToolDescriptor {
     final javaTool
     final javaOptions
 
-    JavaArchiveDescriptor(name, environmentKey, stepConfigurationKey, executablePath, executableName, version, versionOption, javaTool, javaOptions = '') {
-        super(name, environmentKey, stepConfigurationKey, executablePath, executableName, version, versionOption)
+    JavaArchiveDescriptor(name, environmentKey, stepConfigurationKey, version, versionOption, javaTool, javaOptions = '') {
+        super(name, environmentKey, stepConfigurationKey, '', '', version, versionOption)
         this.javaTool = javaTool
         this.javaOptions = javaOptions
     }
 
     @Override
     def getToolLocation(script, configuration, log = true) {
 
-        def home
+        def javaArchiveLocation
+        def javaArchivePath
         if (EnvironmentUtils.isEnvironmentVariable(script, environmentKey)) {
-            home = EnvironmentUtils.getEnvironmentVariable(script, environmentKey)
-            if (log) script.echo "$name home '$home' retrieved from environment."
+            javaArchiveLocation = EnvironmentUtils.getEnvironmentVariable(script, environmentKey)
+            if (log) script.echo "$name location '$javaArchiveLocation' retrieved from environment."
+            javaArchivePath = getJavaArchivePath(javaArchiveLocation)
+            if (javaArchivePath == null) script.error "$environmentKey has an unexpected format."
         }
         else if (configuration.containsKey(stepConfigurationKey)) {
-            home = configuration.get(stepConfigurationKey)
-            if (log) script.echo "$name home '$home' retrieved from configuration."
+            javaArchiveLocation = configuration.get(stepConfigurationKey)
+            if (log) script.echo "$name location '$javaArchiveLocation' retrieved from configuration."
+            javaArchivePath = getJavaArchivePath(javaArchiveLocation)
+            if (javaArchivePath == null) script.error "$stepConfigurationKey has an unexpected format."
         } else if (isOnCurrentWorkingDirectory(script)){
-            home = ''
-            if (log) script.echo "$name expected on current working directory."
+            javaArchivePath = ''
+            if (log) script.echo "$name is on current working directory."
         } else {
             throw new AbortException(getMessage())
         }
-        return home
+        return javaArchivePath
+    }
+
+    def getJavaArchivePath(String javaArchiveLocation) {
```

Comment:
Why do we have the need for distinguishing between the folder containing the jar file and the name of the jar file (including it's extension)?  At the end when we make the java call (something like ```java -jar <mtaUber.jar>) we need only the path to the jar file, something like ```/opt/sap/mta/mta-1.2.3.jar```.

In case the mtaUber jar is expected to be in the current working dir this would also work. In this case we have to configure ```mta-1.2.3.jar```. By the way: this means the customer has to check-in the mtaUber.jar in the sources (at root level). I cannot say that this is a completely stupid use case (... well), but it is a very special approach. But also this does not really (IMO) require to separate the path from the jar file name.

Reply:
It is how it was implemented before, because it was first checking the path that it was configured with mtaJarLocation, and after adding the hard-coded 'mta.jar', and checking then then the full path. The base class continues with this approach, since the home, e.g. NEO_HOME is the value configured by the user, but I can change the implementation once more, should I?

Reply:
But also, I thought that you wanted small changes, if I change the implementation once again, should it not be in a new PR?

Reply:
>because it was first checking the path

With jar files there is no path, like it is for executables (e.g. neo.sh).  The java classpath is something slightly different (searching (compiled) classes inside directories and/or jar files).

>but I can change the implementation once more, should I?

well, it is much noise basically caused by following some assumtions like "but in the base class", "before it was". What we basically would like to check is ```[ -f /opt/java/mta/mta-1.23..jar]```, thats all. I personally would change it. 
&lt;side-remark&gt;By the way: in case we call java -jar my.jar with a non existing jar we get also a clear message from java (in my case: "Error: Unable to access jarfile abc.jar"). But this, of course, depends on the java implement and this is not a general argument against validation, since we may have other tools with less speaking messages.&lt;/side-remark&gt;

>But also, I thought that you wanted small changes,

Here in this case we introduce a new feature (... provide freedom wrt the jar file name). Hence I would put everything we need for that feature into one pull request. We can structure by having reasonable (... small) individual commits. I don't want to continue that discussion into the direction of how github vs gerrit influences the style of the commits ....

Reply:
I guess after 2 years and 4 months we still don't understand each other :)

> With jar files there is no path, like it is for executables (e.g. neo.sh). The java classpath is
> something slightly different (searching (compiled) classes inside directories and/or jar files).
I mean the path to the java archive file, not the PATH.
e.g. if the mtaJarLocation is `/mylocation/mta.jar` --> `/mylocation` is the path to the file

 > well, it is much noise basically caused by following some assumtions like "but in the base
> class", "before it was".
I think that is much noise to change the implementation completely every time that a new feature is implemented.
If there is a base class is for a reason, to reuse code.

> What we basically would like to check is [ -f /opt/java/mta/mta-1.2.3.jar ], thats all. I personally > would change it.
Yes, we only check this.

> Here in this case we introduce a new feature (... provide freedom wrt the jar file name). Hence I > would put everything we need for that feature into one pull request. 
I agree, but we don't need to re-implement the whole class for this feature.

Reply:
Maybe you should look for another colleague to get an approval. I do not vote with "require changes". But I'm not so convinced that I provide "approve". It is not about academical discussions, it has practical impacts. We have to support the code and the less straight-forward it is the harder it is so understand and to support, also for the colleagues after us who have to support it in the future. Applies also for troubleshooting. IMO is is more complicated that needed to split the full path into two parts and make the validation afterwards for each part rather than applying a simple check if the file is present with the configured value "as it" (which is hard enough, as we learned in the meantime, since we have to take docker and slaves into account).

But maybe I'm wrong. Maybe there are some other colleagues who appreciate this approach here.

Reply:
Ok, if you think so I will change it, you have more experience, I don't mind to change it.